File: wiki/deepseek_client.py
"""DeepSeek API client (OpenAI-compatible) with streaming support."""

import os
import json
from typing import AsyncGenerator, Optional
import httpx

DEEPSEEK_BASE_URL = "https://api.deepseek.com/v1"

# ── Hybrid model strategy (cost/quality tiering) ──────────────────────────────
# Knowledge-base construction (Stage 1 distil / Stage 2 assembly / Stage 3 shard)
# needs strong reasoning → v4-pro (quality first).
# Interactive Q&A / page-selection is high-volume → v4-flash (speed/cost first).
# Change models by editing these two constants only.
BUILD_MODEL = "deepseek-v4-pro"
QUERY_MODEL = "deepseek-v4-flash"
# Backwards-compatible default = query tier (matches the iOS port's defaultModel).
DEFAULT_MODEL = QUERY_MODEL
TIMEOUT = 180.0


async def chat_completion(
    messages: list[dict],
    model: str = DEFAULT_MODEL,
    stream: bool = False,
    api_key: Optional[str] = None,
    temperature: float = 0.7,
    timeout: Optional[float] = None,
) -> str | AsyncGenerator[str, None]:
    """
    Call DeepSeek chat completions endpoint.
    stream=True returns an async generator yielding text deltas.
    stream=False returns the full response string.
    API key from DEEPSEEK_API_KEY env var if not passed.
    `timeout` overrides the default per call (large assembly calls need longer).
    """
    key = api_key or os.environ.get("DEEPSEEK_API_KEY", "")
    if not key:
        raise ValueError("DEEPSEEK_API_KEY not set")

    headers = {
        "Authorization": f"Bearer {key}",
        "Content-Type": "application/json",
    }
    payload = {
        "model": model,
        "messages": messages,
        "stream": stream,
        "temperature": temperature,
    }
    eff_timeout = timeout or TIMEOUT

    if stream:
        return _stream_response(headers, payload, eff_timeout)
    else:
        return await _full_response(headers, payload, eff_timeout)


async def _full_response(headers: dict, payload: dict, timeout: float = TIMEOUT) -> str:
    async with httpx.AsyncClient(timeout=timeout) as client:
        response = await client.post(
            f"{DEEPSEEK_BASE_URL}/chat/completions",
            headers=headers,
            json={**payload, "stream": False},
        )
        response.raise_for_status()
        data = response.json()
        return data["choices"][0]["message"]["content"]


async def _stream_response(headers: dict, payload: dict,
                           timeout: float = TIMEOUT) -> AsyncGenerator[str, None]:
    async with httpx.AsyncClient(timeout=timeout) as client:
        async with client.stream(
            "POST",
            f"{DEEPSEEK_BASE_URL}/chat/completions",
            headers=headers,
            json={**payload, "stream": True},
        ) as response:
            response.raise_for_status()
            async for line in response.aiter_lines():
                if not line or not line.startswith("data: "):
                    continue
                data_str = line[6:]
                if data_str == "[DONE]":
                    break
                try:
                    data = json.loads(data_str)
                except json.JSONDecodeError:
                    continue
                if "error" in data:
                    # DeepSeek sends error-shaped SSE events mid-stream for
                    # rate limiting / content filtering instead of raising at
                    # connect time — surface it instead of dropping it silently.
                    err = data["error"]
                    msg = err.get("message", str(err)) if isinstance(err, dict) else str(err)
                    raise RuntimeError(f"DeepSeek 流式响应错误: {msg}")
                try:
                    delta = data["choices"][0]["delta"].get("content", "")
                    if delta:
                        yield delta
                except (KeyError, IndexError):
                    continue

File: app/Reptile.swiftpm/Core/DeepSeekClient.swift
import Foundation

/// DeepSeek API client (OpenAI-compatible), the Swift/URLSession equivalent of
/// wiki/deepseek_client.py.
///
/// 混合模型策略（cost/quality 分层）：
/// - 知识库建设（一级蒸馏 + 二级关系组装）走 `buildModel` = v4-pro：
///   蒸馏与跨文档关系推理需要更强推理，质量优先。
/// - 问答 / 通用对话走 `defaultModel`(=`queryModel`) = v4-flash：
///   交互式、量大，速度与成本优先。
struct DeepSeekClient {
    static let baseURL = "https://api.deepseek.com/v1"
    /// 问答 / 通用对话默认模型（快、省）。
    static let queryModel = "deepseek-v4-flash"
    /// 知识库建设模型（强推理，质量优先）。
    static let buildModel = "deepseek-v4-pro"
    /// 未显式指定时的默认模型 = 问答档。
    static let defaultModel = queryModel
    static let timeout: TimeInterval = 180
    static let assemblyTimeout: TimeInterval = 600

    var apiKey: String

    private func request(path: String, body: [String: Any], timeout: TimeInterval) throws -> URLRequest {
        guard !apiKey.isEmpty else { throw ReptileError.noAPIKey }
        guard let url = URL(string: "\(Self.baseURL)\(path)") else { throw ReptileError.http("bad url") }
        var req = URLRequest(url: url)
        req.httpMethod = "POST"
        req.setValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization")
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.httpBody = try JSONSerialization.data(withJSONObject: body)
        req.timeoutInterval = timeout
        return req
    }

    /// Validate the key via /models (no token cost). Mirrors validate_api_key.
    func validateKey() async -> Bool {
        guard !apiKey.isEmpty, let url = URL(string: "\(Self.baseURL)/models") else { return false }
        var req = URLRequest(url: url)
        req.setValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization")
        req.timeoutInterval = 15
        do {
            let (_, resp) = try await URLSession.shared.data(for: req)
            return (resp as? HTTPURLResponse)?.statusCode == 200
        } catch { return false }
    }

    /// Non-streaming completion → full content string.
    func complete(messages: [[String: String]],
                  model: String = DeepSeekClient.defaultModel,
                  temperature: Double = 0.7,
                  timeout: TimeInterval = DeepSeekClient.timeout) async throws -> String {
        let body: [String: Any] = [
            "model": model, "messages": messages, "stream": false, "temperature": temperature,
        ]
        let req = try request(path: "/chat/completions", body: body, timeout: timeout)
        let (data, resp) = try await URLSession.shared.data(for: req)
        guard let http = resp as? HTTPURLResponse, http.statusCode == 200 else {
            throw ReptileError.http(String(data: data, encoding: .utf8) ?? "HTTP error")
        }
        let json = try JSONSerialization.jsonObject(with: data) as? [String: Any]
        let choices = json?["choices"] as? [[String: Any]]
        let msg = choices?.first?["message"] as? [String: Any]
        return msg?["content"] as? String ?? ""
    }

    /// Streaming completion → async sequence of text deltas (SSE parsing).
    func stream(messages: [[String: String]],
                model: String = DeepSeekClient.defaultModel,
                temperature: Double = 0.7) -> AsyncThrowingStream<String, Error> {
        AsyncThrowingStream { continuation in
            let task = Task {
                do {
                    let body: [String: Any] = [
                        "model": model, "messages": messages, "stream": true, "temperature": temperature,
                    ]
                    let req = try request(path: "/chat/completions", body: body, timeout: Self.timeout)
                    let (bytes, resp) = try await URLSession.shared.bytes(for: req)
                    guard let http = resp as? HTTPURLResponse, http.statusCode == 200 else {
                        throw ReptileError.http("HTTP \((resp as? HTTPURLResponse)?.statusCode ?? -1)")
                    }
                    for try await line in bytes.lines {
                        guard line.hasPrefix("data: ") else { continue }
                        let payload = String(line.dropFirst(6))
                        if payload == "[DONE]" { break }
                        guard let d = payload.data(using: .utf8),
                              let j = try? JSONSerialization.jsonObject(with: d) as? [String: Any]
                        else { continue }
                        if let err = j["error"] as? [String: Any] {
                            let m = err["message"] as? String ?? "stream error"
                            throw ReptileError.http(m)
                        }
                        let choices = j["choices"] as? [[String: Any]]
                        let delta = choices?.first?["delta"] as? [String: Any]
                        if let content = delta?["content"] as? String, !content.isEmpty {
                            continuation.yield(content)
                        }
                    }
                    continuation.finish()
                } catch {
                    continuation.finish(throwing: error)
                }
            }
            continuation.onTermination = { _ in task.cancel() }
        }
    }
}
